### ISSUE TITLE
Solving twice with same VirtualBrownianTree does not produce the same result

### ISSUE THREAD

**Issue by axsk** [SciML/DiffEqNoiseProcess.jl#123]
This seems to have the same problem as #121.

```julia
using StochasticDiffEq, DiffEqBase, DiffEqNoiseProcess, Test, Plots

f1 = (u, p, t) -> 1.01u
g1 = (u, p, t) -> 1.01u
dt = 1 // 2^(4)

W = VirtualBrownianTree(0.0, 0.; tree_depth=6, tend=1.1)
prob1 = SDEProblem(f1, g1, 1.0, (0.0, 1.0), noise=W)
sol1 = solve(prob1, EM(), dt = dt, save_noise = true)
plot(sol1) |> display
prob1 = SDEProblem(f1, g1, 1.0, (0.0, 1.0), noise = W)
sol2 = solve(prob1, EM(), dt = dt)
plot(sol2)
```

Here sol1 and sol2 are not the same, even though the noise should be reproducible.

Is there any other noises where this could also happen?

**Comment by rmsrosa** [comment]
Yeah, same problem. We should definitely check for similar problems for other processes.

**Comment by ChrisRackauckas** [comment]
> Is there any other noises where this could also happen?

There are some tests on doing this with things like NoiseWrapper, and some tests / benchmarks would completely fail if that wasn't the case for NoiseGrid, so I don't think the problem is the solve. Too many other cases like this would be immediately obvious in the benchmark error calculations. 

My guess would be to check the consistency of the VBT. If you solve the NoiseProblem with a given VBT, change the dt around, is it convergent as a noise process? Or does the same VBT give different results when generated twice with the same seed? My guess is the latter. Since it doesn't actually cache values, the VBT needs to be very careful to get this property right.

@frankschae

**Comment by rmsrosa** [comment]
`NoiseProcess`, `NoiseApproximation`, `NoiseWrapper`, `NoiseFunction`, and `NoiseTransport` all have their own proper `reinit!` in https://github.com/SciML/DiffEqNoiseProcess.jl/blob/master/src/noise_interfaces/common.jl.

All others dispatch to `AbstractNoiseProcess`, including `NoiseGrid`, but, `NoiseGrid`, at least, is handled properly, while all others fall short of it, with `W.curW` not resetting.

This should affect not only `VirtualBrownianTree`, but also `SimpleNoiseProcess`, and `BoxWedgeTail`. They need a proper reinitialization of `W.curW` and whatever else.

**Comment by ChrisRackauckas** [comment]
Thanks for that analysis. Looks like we should add a test that just loops through the noise processes and does this first then. SimpleNoiseProcess is a rather straightforward one here to solve, the other two are a bit more complex. Can we start by just getting that test with the `@test_broken` cases highlighted?

I'm actually surprised `SimpleNoiseProcess` wouldn't also be handled by the common.jl one?

**Comment by rmsrosa** [comment]
I started to write the tests but I realized I misunderstood slightly the problem, so I will try to get things straight:

1) The issue in #121 was about `NoiseWrapper` not working properly because it didn't have a `reinit!`. This was fixed. In that situation, whatever noise we use, including VBT, it will be wrapped in `NoiseWrapper` and the `reinit!` for that will take care of it.

2) The processes `VirtualBrownianTree`, `SimpleNoiseProcess` and `BoxWedgeTail` do not have reinit, but that is actually harder to test by itself, because using such noises without being wrapped just means it will use the same process, but not necessarily the same sample path. The problem is there, but it is not as straightforward to write tests as we did for #121.

3) The particular situation described by @axsk is that VBT was setup with `tree_depth=6`, giving a predefined noise at 2^6 + 1 = 65 time instants while the solver was given dt = 1 // 2^(4), hence the noise has a finer mesh. It also had `tend=1.1`, so the mesh points were slightly off, but we could set up `tend=2.0` so that the mesh `W.t` would indeed be a refinement over `sol.t`. In this case, even without using `NoiseWrapper` we would indeed expect the same solution. But this does not happen because `reinit!` is not resetting `W.dW` and that first misstep steers away the second solution.

So, at this point, I don't know how to set up proper tests for the other noises, but I can work on trying to fix this problem for VBT and also checking the reinit for the SimpleNoiseProcess and BoxWedgeTail via debugging.

**Comment by axsk** [comment]
I don't really know what `reinit!` is supposed to do so I cannot comment on that.

However, I thought the idea of a brownian tree was to give a map: rng-state -> sample path. So in particular I would expect `W::VirtualBrownianTree` to hold the rng state and thus always produce the same path.

Also as I understood the `tree_depth` is only for caching and should not influence the noise values in any way.
Even if the `search_depth` is not fine enough this should just lead to constant-noise on the respective intervals, but not hinder deterministic behavior.

Note also that I chose `tend=1.1` because otherwise I got errors for smaller `dt` in which case the solver sometimes solved slightly past `1.0` (I suppose due to roundoff errors).

It might well be though that I misunderstood what VBT is trying to do 😆 .

**Comment by rmsrosa** [comment]
> It might well be though that I misunderstood what VBT is trying to do

You are right. Sorry if I mislead you to think otherwise. `VBT` would have the same sample paths. Just  `SimpleNoiseProcess` and `BoxWedgeTail` would have different sample paths (I am actually not familiar with these, but I am assuming this is the case .. 😅 ).

As for the test, yeah, that is a good test for `VBT`, also. The difficulty is in `SimpleNoiseProcess` and `BoxWedgeTail`, with different sample paths being generated in each solve.

**Comment by rmsrosa** [comment]
Closed by #124
